sklearn/pipeline.py

```diff
@@ -831,3 +841,114 @@ def make_union(*transformers, **kwargs):
                         .format(list(kwargs.keys())[0]))
     return FeatureUnion(_name_estimators(transformers), n_jobs=n_jobs)
+
+
+class ColumnTransformer(FeatureUnion):
+    """Applies transformers to columns of a array / dataframe / dict.
+
+    This estimator applies transformer objects to columns or fields of the
+    input, then concatenates the results. This is useful for heterogeneous or
+    columnar data, to combine several feature extraction mechanisms into a
+    single transformer.
+
+    Read more in the :ref:`User Guide <column_transformer>`.
+
+    Parameters
+    ----------
+    transformers : list of tuples
+        List of (name, transformer, column) tuples specifying the transformer
+        objects to be applied to subsets of the data. The columns can be
+        specified as a scalar or list (for multiple columns) of integer or
+        string values. Integers are interpreted as the positional columns,
+        strings as the keys of `X`. In case of positional integers, also
+        a slice object is accepted.
+
+    n_jobs : int, optional
+        Number of jobs to run in parallel (default 1).
+
+    transformer_weights : dict, optional
+        Multiplicative weights for features per transformer.
+        Keys are transformer names, values the weights.
+
+    Examples
+    --------
+    >>> from sklearn.pipeline import ColumnTransformer
+    >>> from sklearn.preprocessing import Normalizer
+    >>> union = ColumnTransformer(
+    ...     [("norm1", Normalizer(norm='l1'), 'subset1'),
+    ...      ("norm2", Normalizer(norm='l1'), 'subset2')])
+    >>> X = {'subset1': [[0., 1.], [2., 2.]], 'subset2': [[1., 1.], [0., 1.]]}
+    >>> union.fit_transform(X)    # doctest: +NORMALIZE_WHITESPACE
+    array([[ 0. ,  1. ,  0.5,  0.5],
+           [ 0.5,  0.5,  0. ,  1. ]])
+
+    """
+    def _iter(self, X=None, skip_none=True):
+        """Generate (name, trans, column, weight) tuples
+        """
+        get_weight = (self.transformer_weights or {}).get
+        return ((name, trans, _getitem(X, column), get_weight(name))
+                for name, trans, column in self.transformer_list
+                if not skip_none or trans is not None)
+
+    def _update_transformer_list(self, transformers):
+        transformers = iter(transformers)
+        self.transformer_list[:] = [
+            (name, None if old is None else next(transformers), column)
+            for name, old, column in self.transformer_list
+        ]
+
+
+def _getitem(X, column):
+    """
+    Get feature column from input data (array, dataframe, dict)
+
+    """
+    if X is None:
+        return X
+
+    # check whether we have string column names or integers
+    if (isinstance(column, int) or isinstance(column, slice)
+            or (isinstance(column, list)
+                and all(isinstance(col, int) for col in column))):
+        column_names = False
+    elif (isinstance(column, six.string_types)
+          or (isinstance(column, list)
```

Comment:
Yes, indeed. For now I didn't support for simplicity (as I then have to check the type of the slice start and stop), but probably it would be cleaner (and useful) from user perspective to also support this if positional slice is supported. 

But, thinking about it now, the problem becomes: what to do with dicts and recarrays. Raise an error? (for recarrays you could in principle determine what the user wants with some convoluted code, but for a dict not as keys are not sorted)

We could of course also say in general that we don't support getting multiple keys in case of a dict